File: adapters/go/sdk/status.go
package sdk

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"os"
	"regexp"
	"sync"
	"sync/atomic"
	"time"

	"github.com/nats-io/nats.go"
)

// SDKVersion is reported in the status frame so an operator can tell which
// adapters still run an old SDK.
const SDKVersion = "go/0.6.0"

// Status reporting defaults.
const (
	DefaultHeartbeatInterval = 10 * time.Second
	// degradedAfterErrors is when consecutive collect failures stop being
	// noise. Deliberately independent of the device link: a PLC that answers
	// but returns garbage leaves the connection state "connected".
	degradedAfterErrors = 3
	// offlineFlushTimeout bounds the wait for the goodbye frame to reach the
	// server. Short: a shutting-down adapter must not hang on an unreachable
	// broker.
	offlineFlushTimeout = 2 * time.Second
)

// statusReporter publishes adapter runtime status (ADR 0008).
//
// Counters are atomics updated from the collect loop, and the frame is
// assembled by the reporter's own goroutine, so no hook on the hot path does
// more than an atomic add or a non-blocking signal.
type statusReporter struct {
	client *Client
	cfg    *AdapterConfig

	adapterID  string
	instanceID string
	startedAt  time.Time
	interval   time.Duration

	seq                      atomic.Int64
	publishedTotal           atomic.Int64
	collectErrorsTotal       atomic.Int64
	deviceErrorsTotal        atomic.Int64
	deviceReconnectsTotal    atomic.Int64
	consecutiveCollectErrors atomic.Int64

	mu          sync.Mutex
	deviceState DeviceState
	lastError   string
	lastErrorAt *time.Time

	// dirty carries edge-triggered sends. Buffered by one: a burst of
	// transitions between ticks collapses into a single frame rather than
	// queueing, and a full channel is dropped rather than blocking the caller.
	dirty chan struct{}

	stopOnce sync.Once
	stop     chan struct{}
	done     chan struct{}

	// subs are unsubscribed on stop. Without this an adapter that has stopped
	// collecting keeps answering probes with "ok" and re-announcing itself as
	// running, so core reports a dead adapter as healthy.
	subs []*nats.Subscription
}

func newStatusReporter(client *Client, cfg *AdapterConfig) *statusReporter {
	interval := cfg.HeartbeatInterval
	if interval <= 0 {
		interval = DefaultHeartbeatInterval
	}
	adapterID := cfg.AdapterID
	if adapterID == "" {
		// A one-adapter-per-asset deployment needs no new configuration.
		adapterID = cfg.AssetID
	}
	// adapter_id is the last token of the status subject, so it must be a
	// single NATS token. asset_id has no such constraint, and the default
	// falls back to it: an asset named "line3.press" would produce frames core
	// drops with no symptom beyond the adapter never appearing.
	if sanitized := sanitizeAdapterID(adapterID); sanitized != adapterID {
		cfg.Logger.Warn("adapter_id is not a valid NATS subject token; reporting under a sanitized id",
			"from", adapterID, "to", sanitized,
			"hint", "set AdapterID explicitly to control it")
		adapterID = sanitized
	}
	return &statusReporter{
		client:      client,
		cfg:         cfg,
		adapterID:   adapterID,
		instanceID:  newInstanceID(),
		startedAt:   time.Now(),
		interval:    interval,
		deviceState: DeviceDisconnected,
		dirty:       make(chan struct{}, 1),
		stop:        make(chan struct{}),
		done:        make(chan struct{}),
	}
}

// adapterIDInvalid matches every character that cannot appear in a subject
// token. A dot would silently reshape platform.adapter.status.<id> into extra
// tokens; wildcards would make the frame unroutable.
var adapterIDInvalid = regexp.MustCompile(`[^A-Za-z0-9_:\-]`)

// sanitizeAdapterID makes an id usable as a subject token, deterministically so
// that a restart maps to the same id.
func sanitizeAdapterID(id string) string {
	out := adapterIDInvalid.ReplaceAllString(id, "-")
	if out == "" {
		return "adapter"
	}
	if len(out) > 64 {
		out = out[:64]
	}
	// The first character must be alphanumeric.
	if c := out[0]; !(c >= 'a' && c <= 'z' || c >= 'A' && c <= 'Z' || c >= '0' && c <= '9') {
		out = "a" + out
		if len(out) > 64 {
			out = out[:64]
		}
	}
	return out
}

// newInstanceID identifies this process run, so core can tell a restart from a
// continuing process and detect two adapters sharing an id.
func newInstanceID() string {
	buf := make([]byte, 8)
	if _, err := rand.Read(buf); err != nil {
		return "unknown"
	}
	return hex.EncodeToString(buf)
}

// start begins heartbeating and answering hello/ping.
func (r *statusReporter) start(ctx context.Context) {
	if err := r.subscribeControl(); err != nil {
		r.cfg.Logger.Warn("adapter status control subscriptions failed", "err", err)
	}
	r.publish(ctx, AdapterPhaseOnline)
	go r.loop(ctx)
}

func (r *statusReporter) subscribeControl() error {
	nc, err := r.client.conn()
	if err != nil {
		return err
	}
	// hello: core restarted and wants everyone to re-announce.
	hello, err := nc.Subscribe(SubjectAdapterHello, func(_ *nats.Msg) {
		r.publish(context.Background(), AdapterPhaseAnnounce)
	})
	if err != nil {
		return err
	}
	// ping: core missed a heartbeat and is checking before declaring us dead.
	ping, err := nc.Subscribe(SubjectAdapterPingPrefix+r.adapterID, func(msg *nats.Msg) {
		_ = msg.Respond([]byte(`{"ok":true}`))
	})
	if err != nil {
		_ = hello.Unsubscribe()
		return err
	}
	r.subs = append(r.subs, hello, ping)
	return nil
}

func (r *statusReporter) loop(ctx context.Context) {
	defer close(r.done)

	ticker := time.NewTicker(r.interval)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			return
		case <-r.stop:
			return
		case <-ticker.C:
			r.publish(ctx, AdapterPhaseHeartbeat)
		case <-r.dirty:
			// Edge-triggered: a state change is reported immediately rather
			// than waiting up to a whole interval.
			r.publish(ctx, AdapterPhaseHeartbeat)
		}
	}
}

// markDirty requests an out-of-band frame. Non-blocking, so it is safe to call
// from the collect loop and from state transitions.
func (r *statusReporter) markDirty() {
	select {
	case r.dirty <- struct{}{}:
	default:
	}
}

// setDeviceState records a transition. Called after the adapter releases its
// own lock — publishing under that lock would put network I/O inside the state
// machine.
func (r *statusReporter) setDeviceState(s DeviceState) {
	r.mu.Lock()
	changed := r.deviceState != s
	r.deviceState = s
	r.mu.Unlock()
	if changed {
		if s == DeviceError || s == DeviceDisconnected {
			r.deviceErrorsTotal.Add(1)
		}
		if s == DeviceReconnecting {
			r.deviceReconnectsTotal.Add(1)
		}
		r.markDirty()
	}
}

func (r *statusReporter) incPublished() {
	r.publishedTotal.Add(1)
	r.consecutiveCollectErrors.Store(0)
}

func (r *statusReporter) incCollectError(err error) {
	r.collectErrorsTotal.Add(1)
	n := r.consecutiveCollectErrors.Add(1)

	r.mu.Lock()
	now := time.Now()
	r.lastError = err.Error()
	r.lastErrorAt = &now
	r.mu.Unlock()

	// Report the moment the adapter becomes degraded rather than waiting for
	// the next tick; that transition is the whole point of the axis.
	if n == degradedAfterErrors {
		r.markDirty()
	}
}

func (r *statusReporter) runState() RunState {
	if r.consecutiveCollectErrors.Load() >= degradedAfterErrors {
		return RunStateDegraded
	}
	return RunStateRunning
}

// frame assembles the current status.
func (r *statusReporter) frame(phase string) AdapterStatusFrame {
	r.mu.Lock()
	deviceState := r.deviceState
	lastError := r.lastError
	lastErrorAt := r.lastErrorAt
	r.mu.Unlock()

	now := time.Now()
	runState := r.runState()
	if phase == AdapterPhaseOffline {
		runState = RunStateStopped
	}

	counters := AdapterCounters{
		PublishedTotal:           r.publishedTotal.Load(),
		CollectErrorsTotal:       r.collectErrorsTotal.Load(),
		DeviceErrorsTotal:        r.deviceErrorsTotal.Load(),
		DeviceReconnectsTotal:    r.deviceReconnectsTotal.Load(),
		ConsecutiveCollectErrors: r.consecutiveCollectErrors.Load(),
	}

	assets := []AdapterAssetStatus{{
		AssetID:            r.cfg.AssetID,
		DeviceState:        string(deviceState),
		PublishedTotal:     counters.PublishedTotal,
		CollectErrorsTotal: counters.CollectErrorsTotal,
		LastError:          lastError,
		LastErrorAt:        lastErrorAt,
	}}

	frame := AdapterStatusFrame{
		SchemaVersion:      AdapterSchemaVersion,
		AdapterID:          r.adapterID,
		InstanceID:         r.instanceID,
		Seq:                r.seq.Add(1),
		Phase:              phase,
		RunState:           runState,
		DeviceState:        string(deviceState),
		HeartbeatIntervalS: int(r.interval / time.Second),
		UptimeS:            int64(now.Sub(r.startedAt).Seconds()),
		StartedAt:          r.startedAt.UTC(),
		SentAt:             now.UTC(),
		SDK:                SDKVersion,
		AdapterVersion:     r.cfg.AdapterVersion,
		Capabilities:       []string{"ping"},
		Assets:             assets,
		DeviceCounts:       map[string]int{string(deviceState): 1},
		Counters:           counters,
	}
	if r.cfg.ReportHost {
		frame.Host, _ = os.Hostname()
		frame.PID = os.Getpid()
	}
	return frame
}

func (r *statusReporter) publish(ctx context.Context, phase string) {
	raw, err := json.Marshal(r.frame(phase))
	if err != nil {
		r.cfg.Logger.Warn("encode adapter status", "err", err)
		return
	}
	if err := r.client.PublishRaw(ctx, SubjectAdapterStatusPrefix+r.adapterID, raw); err != nil {
		// Status is best-effort by design: failing to report must never take
		// down an adapter that is otherwise collecting fine.
		r.cfg.Logger.Debug("publish adapter status", "err", err)
	}
}

// stopWith sends a final frame and halts. Called from a deferred function
// registered after the client's own close defer, so LIFO ordering guarantees
// the connection is still open here.
func (r *statusReporter) stopWith(ctx context.Context) {
	r.stopOnce.Do(func() {
		close(r.stop)
		<-r.done
		// Stop answering before saying goodbye: a probe answered after the
		// offline frame would resurrect a stopped adapter in core's registry.
		for _, sub := range r.subs {
			_ = sub.Unsubscribe()
		}
		r.subs = nil
		r.publish(ctx, AdapterPhaseOffline)
		// Publishing is buffered and Client.Close drains asynchronously, so
		// without an explicit flush the goodbye frame is lost whenever the
		// process exits promptly after Run returns — which is the normal case.
		// Core would then report the adapter stale minutes later instead of
		// knowing at once that it stopped cleanly.
		if err := r.client.Flush(offlineFlushTimeout); err != nil {
			r.cfg.Logger.Debug("flush adapter offline frame", "err", err)
		}
	})
}

File: adapters/python/sdk/status.py
"""Adapter runtime status reporting (ADR 0008).

Mirrors adapters/go/sdk/status.go field for field. The two SDKs share a golden
JSON fixture so a field that drifts on either side fails a test rather than
silently producing frames core cannot read.
"""

from __future__ import annotations

import asyncio
import json
import logging
import os
import re
import secrets
import socket
import time
from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import Enum
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .client import NATSClientWrapper

logger = logging.getLogger(__name__)

SDK_VERSION = "python/0.6.0"
ADAPTER_SCHEMA_VERSION = 1

# Subjects. Mirrors internal/core/events.go and the Go SDK's subjects.go.
SUBJECT_ADAPTER_STATUS_PREFIX = "platform.adapter.status."
SUBJECT_ADAPTER_HELLO = "platform.adapter.hello"
SUBJECT_ADAPTER_PING_PREFIX = "platform.adapter.ping."

DEFAULT_HEARTBEAT_INTERVAL = 10.0

# Consecutive collect failures before the adapter calls itself degraded.
# Independent of the device link on purpose: a PLC that answers but returns
# garbage leaves the connection state "connected".
DEGRADED_AFTER_ERRORS = 3

# Bounds the wait for the goodbye frame to reach the server. Short: a shutting
# down adapter must not hang on an unreachable broker.
OFFLINE_FLUSH_TIMEOUT = 2


# Every character that cannot appear in a subject token. A dot would silently
# reshape platform.adapter.status.<id> into extra tokens.
_ADAPTER_ID_INVALID = re.compile(r"[^A-Za-z0-9_:\-]")


def sanitize_adapter_id(adapter_id: str) -> str:
    """Make an id usable as a subject token, deterministically so a restart
    maps to the same id."""
    out = _ADAPTER_ID_INVALID.sub("-", adapter_id)
    if not out:
        return "adapter"
    out = out[:64]
    if not out[0].isalnum():
        out = ("a" + out)[:64]
    return out


class RunState(str, Enum):
    """Adapter process lifecycle, distinct from the device link."""

    STARTING = "starting"
    RUNNING = "running"
    DEGRADED = "degraded"
    STOPPING = "stopping"
    STOPPED = "stopped"


class AdapterPhase(str, Enum):
    """Why a frame was sent. Lets core tell a clean shutdown from silence."""

    ONLINE = "online"
    HEARTBEAT = "heartbeat"
    ANNOUNCE = "announce"
    PROBE = "probe"
    OFFLINE = "offline"


@dataclass
class AdapterCounters:
    """Monotonic totals. Core differences consecutive observations."""

    published_total: int = 0
    collect_errors_total: int = 0
    device_errors_total: int = 0
    device_reconnects_total: int = 0
    nats_reconnects_total: int = 0
    consecutive_collect_errors: int = 0

    def to_dict(self) -> dict:
        return {
            "published_total": self.published_total,
            "collect_errors_total": self.collect_errors_total,
            "device_errors_total": self.device_errors_total,
            "device_reconnects_total": self.device_reconnects_total,
            "nats_reconnects_total": self.nats_reconnects_total,
            "consecutive_collect_errors": self.consecutive_collect_errors,
        }


@dataclass
class StatusReporter:
    """Publishes adapter runtime status on a background asyncio task."""

    client: "NATSClientWrapper"
    asset_id: str
    adapter_id: str = ""
    adapter_version: str = ""
    heartbeat_interval: float = DEFAULT_HEARTBEAT_INTERVAL
    report_host: bool = False

    counters: AdapterCounters = field(default_factory=AdapterCounters)
    _seq: int = 0
    _instance_id: str = ""
    _started_at: float = 0.0
    _device_state: str = "disconnected"
    _last_error: str = ""
    _last_error_at: str | None = None
    _task: asyncio.Task | None = None
    _subs: list = field(default_factory=list)
    _dirty: asyncio.Event | None = None
    _running: bool = False

    def __post_init__(self) -> None:
        if not self.adapter_id:
            # One adapter per asset needs no new configuration.
            self.adapter_id = self.asset_id
        # adapter_id is the last token of the status subject, so it must be a
        # single NATS token. asset_id has no such constraint and the default
        # falls back to it: an asset named "line3.press" would produce frames
        # core drops, with no symptom beyond the adapter never appearing.
        sanitized = sanitize_adapter_id(self.adapter_id)
        if sanitized != self.adapter_id:
            logger.warning(
                "adapter_id %r is not a valid NATS subject token; reporting as %r. "
                "Set adapter_id explicitly to control it.",
                self.adapter_id,
                sanitized,
            )
            self.adapter_id = sanitized
        if self.heartbeat_interval <= 0:
            self.heartbeat_interval = DEFAULT_HEARTBEAT_INTERVAL
        # Identifies this process run so core can detect restarts and two
        # adapters sharing an id.
        self._instance_id = secrets.token_hex(8)
        self._started_at = time.time()

    async def start(self) -> None:
        """Publish the online frame, subscribe to hello/ping and begin
        heartbeating."""
        self._dirty = asyncio.Event()
        self._running = True
        await self._subscribe_control()
        await self.publish(AdapterPhase.ONLINE)
        self._task = asyncio.create_task(self._loop())

    async def stop(self) -> None:
        """Send the goodbye frame and halt.

        Must be awaited before the NATS connection closes, which is why
        BaseAdapter.stop calls it ahead of client.disconnect().
        """
        self._running = False
        if self._dirty is not None:
            self._dirty.set()
        if self._task is not None:
            self._task.cancel()
            try:
                await self._task
            except asyncio.CancelledError:
                pass
            self._task = None

        # Stop answering before saying goodbye: a probe answered after the
        # offline frame would resurrect a stopped adapter in core's registry.
        for sub in self._subs:
            try:
                await sub.unsubscribe()
            except Exception as e:  # pragma: no cover - defensive
                logger.debug(f"Unsubscribe adapter control failed: {e}")
        self._subs = []

        await self.publish(AdapterPhase.OFFLINE)

        # Publishing is buffered and disconnect() drains asynchronously, so
        # without an explicit flush the goodbye frame is lost whenever the
        # process exits promptly. Core would then report the adapter stale
        # minutes later instead of knowing at once that it stopped cleanly.
        nc = self.client.nc
        if nc is not None:
            try:
                await nc.flush(timeout=OFFLINE_FLUSH_TIMEOUT)
            except Exception as e:
                logger.debug(f"Flush adapter offline frame failed: {e}")

    async def _subscribe_control(self) -> None:
        nc = self.client.nc
        if nc is None:
            return
        try:
            # hello: core restarted and wants everyone to re-announce.
            self._subs.append(await nc.subscribe(
                SUBJECT_ADAPTER_HELLO,
                cb=lambda _msg: asyncio.create_task(self.publish(AdapterPhase.ANNOUNCE)),
            ))
            # ping: core missed a heartbeat and checks before declaring us dead.
            self._subs.append(await nc.subscribe(
                SUBJECT_ADAPTER_PING_PREFIX + self.adapter_id,
                cb=self._handle_ping,
            ))
        except Exception as e:  # pragma: no cover - defensive
            logger.warning(f"Adapter status control subscriptions failed: {e}")

    async def _handle_ping(self, msg) -> None:
        if msg.reply:
            await self.client.nc.publish(msg.reply, b'{"ok":true}')

    async def _loop(self) -> None:
        while self._running:
            try:
                # Wake on either the heartbeat tick or an edge-triggered
                # transition, so a state change is reported at once instead of
                # up to a whole interval later.
                await asyncio.wait_for(self._dirty.wait(), timeout=self.heartbeat_interval)
                self._dirty.clear()
            except asyncio.TimeoutError:
                pass
            except asyncio.CancelledError:
                return
            if not self._running:
                return
            await self.publish(AdapterPhase.HEARTBEAT)

    def mark_dirty(self) -> None:
        """Request an out-of-band frame. Safe to call from the collect loop."""
        if self._dirty is not None:
            self._dirty.set()

    def set_device_state(self, state: str) -> None:
        changed = self._device_state != state
        self._device_state = state
        if not changed:
            return
        if state in ("error", "disconnected"):
            self.counters.device_errors_total += 1
        if state == "reconnecting":
            self.counters.device_reconnects_total += 1
        self.mark_dirty()

    def inc_published(self) -> None:
        self.counters.published_total += 1
        self.counters.consecutive_collect_errors = 0

    def inc_collect_error(self, error: Exception) -> None:
        self.counters.collect_errors_total += 1
        self.counters.consecutive_collect_errors += 1
        self._last_error = str(error)
        self._last_error_at = _now_iso()
        # Report the moment the adapter becomes degraded; that transition is
        # the whole point of the axis.
        if self.counters.consecutive_collect_errors == DEGRADED_AFTER_ERRORS:
            self.mark_dirty()

    def run_state(self) -> RunState:
        if self.counters.consecutive_collect_errors >= DEGRADED_AFTER_ERRORS:
            return RunState.DEGRADED
        return RunState.RUNNING

    def frame(self, phase: AdapterPhase) -> dict:
        """Build a status frame. Field names and types mirror the Go SDK."""
        self._seq += 1
        now = time.time()
        run_state = RunState.STOPPED if phase == AdapterPhase.OFFLINE else self.run_state()

        asset: dict = {
            "asset_id": self.asset_id,
            "device_state": self._device_state,
            "published_total": self.counters.published_total,
            "collect_errors_total": self.counters.collect_errors_total,
        }
        if self._last_error:
            asset["last_error"] = self._last_error
            asset["last_error_at"] = self._last_error_at

        frame: dict = {
            "schema_version": ADAPTER_SCHEMA_VERSION,
            "adapter_id": self.adapter_id,
            "instance_id": self._instance_id,
            "seq": self._seq,
            "phase": phase.value,
            "run_state": run_state.value,
            "device_state": self._device_state,
            "heartbeat_interval_s": int(self.heartbeat_interval),
            "uptime_s": int(now - self._started_at),
            "started_at": _iso(self._started_at),
            # Diagnostic only: core judges liveness by its own receive time, so
            # a wrong clock here cannot make a live adapter look dead.
            "sent_at": _now_iso(),
            "sdk": SDK_VERSION,
            "capabilities": ["ping"],
            "config_version": 0,
            "assets": [asset],
            "device_counts": {self._device_state: 1},
            "counters": self.counters.to_dict(),
        }
        if self.adapter_version:
            frame["adapter_version"] = self.adapter_version
        if self.report_host:
            frame["host"] = socket.gethostname()
            frame["pid"] = os.getpid()
        return frame

    async def publish(self, phase: AdapterPhase) -> None:
        nc = self.client.nc
        if nc is None:
            return
        try:
            payload = json.dumps(self.frame(phase)).encode()
            await nc.publish(SUBJECT_ADAPTER_STATUS_PREFIX + self.adapter_id, payload)
        except Exception as e:
            # Status is best-effort by design: failing to report must never
            # take down an adapter that is otherwise collecting fine.
            logger.debug(f"Publish adapter status failed: {e}")


def _iso(epoch: float) -> str:
    return datetime.fromtimestamp(epoch, tz=timezone.utc).isoformat().replace("+00:00", "Z")


def _now_iso() -> str:
    return _iso(time.time())
